Use System.Text.Json instead of SimpleJson

louis-z opened the pull request:
This entails the following:
- Add specific type converters for `enum`s, `DynamicEnum<T>`, `Sha1`, etc.
- Adapt API so that no JSON serialization types are exposed (as was the case for `GitLabException.ErrorObject`)
 
## Rationale for changing serializers
`System.Text.Json` is a modern serializer that offers much better performances than `SimpleJson`. For instance, using a simple benchmark that deserializes an array of 20 `Project`s, we get the following performances:

![image](https://user-images.githubusercontent.com/38251605/150783997-016cde34-50f1-4b08-89ca-b4e1a845601b.png)

`System.Text.Json` is 5 times faster and allocates 100 times less memory.

louis-z committed:
Use System.Text.Json instead of SimpleJson

This entails the following:
- Add specific type converters for enums, DynamicEnum<T> and Sha1
- Adapt API so that
  - value types that could be deserialized from null are defined as nullable
  - no JSON serialization types are exposed (as was the case for GitLabException.ErrorObject)

meziantou commented on NGitLab.Tests/BranchClientTests.cs:
```diff
@@ -26,7 +26,7 @@ public async Task Test_CommitInfoIsCorrectlyDeserialized()
 
             var fiveMinutesAgo = DateTime.UtcNow - TimeSpan.FromMinutes(5);
-            Assert.Less(fiveMinutesAgo, commit.CreatedAt);
+            Assert.Less(fiveMinutesAgo, commit.CreatedAt.ToUniversalTime());
```
The current version doesn't return UTC date (`DateTime.Kind`)? Maybe we should fix that in another MR if needed

louis-z replied:
IIRC the result of deserializing `"created_at"` differs
- `SimpleJson` converts it to UTC, while
- `System.Text.Json` keeps it in Local time

meziantou replied:
GitLab is inconsistent (once again). For an unknown reason it uses "+00:00" instead of "Z" for `created_at`...
So, S.T.Json converts it to a local date.

I would prefer a custom converter if needed to avoid potential issues with date comparison.

The test should be unchanged.

louis-z replied:
Added a custom converter & put back test as it was before.

meziantou commented on NGitLab/GitLabException.cs:
```diff
@@ -37,8 +37,6 @@ public GitLabException(string message, Exception inner)
         /// </summary>
         public HttpStatusCode StatusCode { get; set; }
 
-        public JsonObject ErrorObject { get; set; }
```
I think you can preserve this property (but changing the return type) by using `JsonExtensionDataAttribute`

louis-z replied:
I always felt it was a mistake to leak out to the public API types that are associated to the serializer. Anything related to the serializer should remain internal.

Furthermore, has `ErrorObject` ever been useful to anyone? I think not, but if I'm wrong, I propose that we make it something more generic. Maybe `IDictionary<string, object>`?

meziantou replied:
We do use this property to extract error details in some cases.
`IDictionary<string, object>` is fine

louis-z replied:
Done

meziantou commented on NGitLab/Impl/Json/EnumConverter.cs:
```diff
@@ -0,0 +1,79 @@
+        public override bool CanConvert(Type type)
+        {
+            // AccessLevel will be serialized as int, using System.Text.Json's default enum serializer
+            return type.IsEnum && type != typeof(AccessLevel);
```
Can't you use `JsonStringEnumConverter`?

louis-z replied:
AFAIK `JsonStringEnumConverter` does not handle the `[EnumMember]` attribute.

louis-z committed:
Remove System.Text.Json stuff from FileDelete & FileUpsert

It turns out FileDelete & FileUpsert are actually transfered as query string parameters, not json.

louis-z committed:
Delete SimpleJson.cs & rename SystemTextJsonSerializer to Serializer

meziantou commented on NGitLab/Impl/API.cs:
```diff
@@ -64,7 +64,6 @@ private string OpenPrivateSession()
                 {
                     OriginalCall = new Uri(ex.OriginalCall.OriginalString.Replace(_credentials.Password, hiddenPassword)),
                     StatusCode = ex.StatusCode,
-                    ErrorObject = ex.ErrorObject,
```
Keep the property and use `IDictionary<string, object?>`

meziantou commented on NGitLab/Impl/Json/ReflectionExtensions.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+    }
+
+    internal struct EnumMapping
```
```suggestion
    internal readonly struct EnumMapping
```

meziantou commented on NGitLab/Impl/Json/Serializer.cs:
```diff
@@ -0,0 +1,49 @@
+                obj = Deserialize<object>(json);
+                return true;
+            }
+            catch (Exception /*ex*/)
```
```suggestion
            catch
```

meziantou commented on NGitLab/Models/MergeRequestApprove.cs:
```diff
@@ -9,12 +10,14 @@ public class MergeRequestApprove
         /// (optional) - Current user’s password. Required if Require user password to approve is enabled in the project settings.
         /// </summary>
         [DataMember(Name = "approval_password ")]
+        [JsonPropertyName("approval_password ")]
```
```suggestion
        [JsonPropertyName("approval_password")]
```

meziantou commented on NGitLab/PublicAPI.Unshipped.txt:
```diff
@@ -2818,6 +2793,7 @@ NGitLab.Models.UserToken.CreatedAt.get -> System.DateTime
 NGitLab.Models.UserToken.CreatedAt.set -> void
 NGitLab.Models.UserToken.ExpiresAt.get -> System.DateTime?
 NGitLab.Models.UserToken.ExpiresAt.set -> void
+NGitLab.Models.UserToken.ExpiresAtStr.get -> string
```
I think we can avoid these new props by using a custom converter for DateTime

louis-z replied:
Done

meziantou commented on NGitLab/PublicAPI.Unshipped.txt:
```diff
@@ -2773,22 +2748,22 @@ NGitLab.Models.User.CanCreateGroup -> bool
+NGitLab.Models.User.ConfirmedAt -> System.DateTime?
 NGitLab.Models.User.CreatedAt -> System.DateTime
-NGitLab.Models.User.CurrentSignIn -> System.DateTime
+NGitLab.Models.User.CurrentSignIn -> System.DateTime?
```
Could we implement a custom converter to set the prop to `DateTime.MinValue` (current behavior) when the json prop is null?

meziantou commented on NGitLab/PublicAPI.Unshipped.txt:
```diff
@@ -2550,14 +2525,14 @@ NGitLab.Models.ResourceLabelEventAction.Add = 0 -> NGitLab.Models.ResourceLabelE
 NGitLab.Models.Runner.Locked -> bool
 NGitLab.Models.Runner.Name -> string
-NGitLab.Models.Runner.Online -> bool
+NGitLab.Models.Runner.Online -> bool?
```
Could we use a custom converter to avoid the breaking change?

meziantou commented on NGitLab/PublicAPI.Unshipped.txt:
```diff
@@ -1358,7 +1333,7 @@ NGitLab.Models.Event.TargetIId.get -> long?
 NGitLab.Models.Event.TargetTitle.get -> string
 NGitLab.Models.Event.TargetTitle.set -> void
-NGitLab.Models.Event.TargetType.get -> NGitLab.DynamicEnum<NGitLab.Models.EventTargetType>
+NGitLab.Models.Event.TargetType.get -> NGitLab.DynamicEnum<NGitLab.Models.EventTargetType>?
```
I think a custom converter could map `null` to `EventTargetType.None` to avoid the breaking change

louis-z replied:
Done

louis-z committed:
fixup

meziantou commented on NGitLab/Impl/Json/BooleanConverter.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace NGitLab.Impl.Json
+{
+    internal sealed class BooleanConverter : JsonConverter<bool>
```
Could you add some tests for these converters?

louis-z replied:
The converters were added to make existing tests pass. So, the converter code is already validated (and without it tests would fail). Do you still think we need to add more tests?

meziantou replied:
You don't need to validate them all, but I would prefer some tests to validate it correctly handle `null` values and also the parsing of datetime in different format (and that they are all UTC).

louis-z replied:
I added some tests. Let me know if you feel there should be more.

meziantou commented on NGitLab/Impl/Json/Int64Converter.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace NGitLab.Impl.Json
+{
+    internal sealed class Int64Converter : JsonConverter<long>
```
Could you add a test to validate this converter is not used for `long?`?

louis-z replied:
According to the [System.Text.Json documentation](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to?pivots=dotnet-6-0#handle-null-values), this custom converter will **not** be used for `Nullable<long>`:

```
By default, the serializer handles null values as follows:
- For reference types and Nullable<T> types:
  - It does not pass null to custom converters on serialization.
  - It does not pass JsonTokenType.Null to custom converters on deserialization.
  - It returns a null instance on deserialization.
  - It writes null directly with the writer on serialization.
- For non-nullable value types:
  - It passes JsonTokenType.Null to custom converters on deserialization. (If no custom converter is available, a JsonException exception is thrown by the internal converter for the type.)
```

So unless you think the documentation is lying 😁, I don't see the need for such a test.

What do you think?

meziantou replied:
Let's trust the documentation!

meziantou commented on NGitLab/Impl/Json/Int64Converter.cs:
```diff
@@ -0,0 +1,33 @@
+            if (reader.TokenType == JsonTokenType.String)
+            {
+                var stringValue = reader.GetString();
+                if (long.TryParse(stringValue, out var value))
```
CultureInvariant

louis-z replied:
Done

louis-z committed:
fixup

louis-z committed:
Add requested tests

meziantou commented on NGitLab.Tests/Impl/JsonConverterTests.cs:
```diff
@@ -0,0 +1,88 @@
+            var json = $@"{{ ""a_date_time"": ""{input}"" }}";
+            var obj = Serializer.Deserialize<MyDataContract>(json);
+            Assert.NotNull(obj);
+            Assert.AreEqual(DateTimeKind.Utc, obj.SomeDateTime.Kind);
```
Should it be unknown for `[TestCase("2022-01-12")]`?

louis-z replied:
Yeah, I guess that would make sense. I'll revisit the code...

louis-z replied:
Done.

Also, I removed unnecessary `DateTime` format strings, i.e. those without which the tests still pass.

louis-z committed:
Address PR comments
